[TI-4422] Support files larger than 4 GiB

tkinz27 opened the pull request:
Linter got a little aggressive - core change is ensuring offset calculations are done using `uint64_t` size objects.

Tested (jointly with https://github.com/braincorp/squashmerge/pull/4):
- [x] Backwards compatibility with unchanged squashmerge: No, it doesn't recognize the file.
- [x] squashfs target containing a single large file (33 GiB): Pass
- [x] squashfs target containing a max-length symlink (255 chars possible on ext4): Pass
- [x] ~~squashfs target containing a directory with many files:~~ No need. `/var/lib/dpkg/info` already has 5374 files in it.

johngill-bc committed:
Attempt 2

johngill-bc commented on src/squashdelta.cxx:
```diff
@@ -56,31 +56,32 @@ struct sqdelta_header
 };
 #pragma pack(pop)
 
+#define htonl(val) val
```
Need to figure out how to support this function for 64bit

johngill-bc replied:
Added a new function for handling 64bit

johngill-bc commented on src/squashdelta.cxx:
```diff
@@ -42,7 +42,7 @@ struct compressed_block
 struct serialized_compressed_block
 {
-	uint32_t offset;
+	uint64_t offset;
```
Core change here.

johngill-bc commented on src/squashdelta.cxx:
```diff
@@ -130,14 +132,13 @@ std::list<struct compressed_block> get_blocks(MMAPFile& f, Compressor*& c,
-			uint32_t pos;
-			uint32_t block_count;
-			le32* block_list;
+			uint64_t pos;
+			uint64_t block_count;
```
This is other core change

johngill-bc committed:
Handle endianness

jordisolanas-bc commented on src/squashdelta.cxx:
```diff
@@ -157,8 +174,7 @@ std::list<struct compressed_block> get_blocks(MMAPFile& f, Compressor*& c,
 					// seek over the uncompressed block
-					uint32_t len = (block_list[j]
-							& ~squashfs::block_size::uncompressed);
+					uint32_t len = (block_list[j] & ~squashfs::block_size::uncompressed);
```
In this loop (3 lines above), if `block_count` is now `uint64_t`. should we declare the iterator `j` as `uint64_t` too?
Otherwise `for (uint32_t j = 0; j < block_count; ++j)` could overflow `j` and potentially cause an infinite loop.

davidhasegawa-bc committed:
Make for loop iterator as big as the limit

davidhasegawa-bc committed:
More uint64_t

davidhasegawa-bc committed:
Could size_t be too small?

davidhasegawa-bc commented on src/squashdelta.cxx:
```diff
@@ -33,16 +33,16 @@ extern "C"
-	size_t offset;
-	size_t length;
-	size_t uncompressed_length;
+	uint64_t offset;
+	uint32_t length;
+	uint32_t uncompressed_length;
```
Does this change make it more correct, if we were to process a delta on a 32-bit system for some reason (who knows if we try to optimize robot costs)? Or would the rest of the program not work anyway on such a system?

tkinz27 replied:
I'm not 100% sure I understand how size_t gets set. We are already compiling on 64bit systems. IIRC the reason these are explicitly set now is that we have a mix of uint32_t and uint64_t below.

jordisolanas-bc replied:
That's a tricky one, but I would say this is more "correct".
While size of `size_t` changes based on system/compiler (usually 32 bits in a 32-bits system,  and 64 in a 64-bits system), compilers nowadays handle uint64_t in 32-bits environments. While performance is slower, it allows you to represent a 64 bit data (which would still be necessary to process squashfs images > 4GB).

davidhasegawa-bc replied:
After thinking and reading some more of the code, I don't think this program would work with large files on a 32-bit system, because it reads the source squashfs via a memory map.

davidhasegawa-bc replied:
To clarify Tony, I added this change right now and am asking if it's correct. (Because you can't comment on a line on GitHub without changing it.)

davidhasegawa-bc replied:
Now I'm thinking `size_t` is more correct, because these are representing offsets and lengths in the memory map.

davidhasegawa-bc committed:
Add comments on endianness

davidhasegawa-bc committed:
Undo type changes for regular file inodes

The squashfs format specifies they are 32-bit.
https://dr-emann.github.io/squashfs/squashfs.html#_inode_table

davidhasegawa-bc committed:
Update usage of lreg::block_count's return

davidhasegawa-bc committed:
Revert "Could size_t be too small?"

This reverts commit b681feab

davidhasegawa-bc committed:
Add comments

davidhasegawa-bc committed:
Ignore JetBrains IDE dir

davidhasegawa-bc committed:
Allow assigning from uint64_t to le64

davidhasegawa-bc committed:
Increase buf_size to max possible inode size

davidhasegawa-bc committed:
Move buffer from stack to heap

davidhasegawa-bc commented on src/squashfs.hxx:
```diff
@@ -266,9 +276,11 @@ public:
-	static const size_t buf_size = 16 * squashfs::metadata_size;
-	char buf[buf_size];
+
+	static constexpr size_t max_supported_file_size = 1ull << 40; // 1 TiB -> 1 GiB buffer (2^30)
```
Note, this doesn't take up 1 GiB of physical memory unless that part of the buffer is actually written to.
https://stackoverflow.com/questions/864416/are-some-allocators-lazy

davidhasegawa-bc replied:
Wrote this comment in the code.

jordisolanas-bc replied:
Could you please clarify what the comment `// 1 TiB -> 1 GiB buffer (2^30)` means? Not sure I understand what you mean with it?

davidhasegawa-bc replied:
The max file size we're declaring we support is 2^40 (1 TiB), which results in a max inode size of 2^30 (1 GiB) plus a few bytes. That means our buffer needs to be at least that large to read the inode.

1 TiB is arbitrary, but seemed future-proof since that's the size of our entire SSD.

jordisolanas-bc replied:
Can you please clarify the comment in code?

davidhasegawa-bc replied:
Done!

davidhasegawa-bc committed:
Explain max_supported_file_size in comment

davidhasegawa-bc committed:
Make max_possible_inode_size calculation more readable

jordisolanas-bc commented on src/squashfs.hxx:
```diff
@@ -198,8 +203,14 @@ namespace squashfs
+				// file (2^64) and smallest possible block size (2^12).
+				// max_file_size / 2^12 * sizeof(le32) == max_file_size >> 10
+                return sizeof(lreg) + (max_file_size >> 10);
+            }
```
Minor: Indent seems to be wrong here

jordisolanas-bc commented on src/squashfs.hxx:
```diff
@@ -169,6 +170,10 @@ namespace squashfs
 			size_t inode_size();
+			static constexpr size_t max_possible_inode_size(uint_fast32_t max_symlink_target_size) {
+                // Max possible symlink happens with longest possible symlink target (2^32)
+                return sizeof(symlink) + max_symlink_target_size;
```
Minor: indent

jordisolanas-bc commented on src/squashfs.cxx:
```diff
@@ -149,6 +149,12 @@ MetadataReader::MetadataReader(const MMAPFile& new_file,
 	: f(new_file, offset, c),
 	bufp(buf), buf_filled(0), _block_num(0)
 {
+	buf = new char[buf_size];
```
Mmm... Before the changes, this initialization was causing `bufp` to point to `buf` (which was statically allocated). Now `buf` is dinamically allocated after `bufp(buf)`, not sure this is what you intended here.

davidhasegawa-bc replied:
Good catch, they're supposed to start off pointing to the same spot.

davidhasegawa-bc committed:
Explain choosing of max_supported_file_size

davidhasegawa-bc committed:
Fix initialization of bufp

davidhasegawa-bc committed:
Fix spaces vs. tabs
